### ISSUE TITLE
[feature request] Add option for scrolling window management

### ISSUE THREAD

**Issue by Lillian-Violet** [pop-os/cosmic-epoch#322]
While the tiling window management style is tried and tested, and quite good for a lot of use cases. I would like to see an option for scrolling window management like the [cardboard wm POC](https://github.com/cardboardwm/cardboard?tab=readme-ov-file). I have seen a lot of enthousiasm for this concept, and I think a lot of people would adopt it given the chance. But it hasn't really been implemented in a big project in a satisfactory way.

The main concepts are:
1. Basics are like a normal tile window manager; you have windows that open to a snapped grid, you can resize and reposition the windows.
2. Instead of new windows resizing old windows to fit on the screen, they open to the side off-screen.
3. You can change what application you are looking at by scrolling through your tiles (with three fingers on a touchpad, holding specific keys etc.) to change what applications are in the center of your screen.

**Comment by Drakulix** [comment]
We will probably not introduce more window management concepts at this point in development.

But there have been frequent requests for more options, so we might (long-term) try to provide an api to have external programs add new window management options.
Additionally we want to enable changing the compositor, so you could e.g. run niri instead of cosmic-comp in your session to get scrolling window management.

**Comment by boggle** [comment]
The lack of paperwm style scrollable tiling is what is keeping me on gnome for the time being. I believe reasonable support for this could be hidden behind a single tick box.

**Comment by mongaikan** [comment]
While waiting for COSMIC I tried PaperWM in Gnome and as a long time regular manual and auto tiling WMs user I must say that I grew fond of that workflow as well.
It makes sense to me, you almost never have to think about window management anymore. I get stuff done, it doesn't get in my way.

I know I'll be switching to COSMIC as soon as possible for me anyway but adding a native scrollable window management would make COSMIC THE ultimate DE it's meant to be in my opinion.
I don't know any existing window management concept it wouldn't cover then?

I'm sure it's not a trivial endeavor considering all the care and sophisticated tiling management that's already defining COSMIC as the best tiler out there despite it being an alpha at the time of writing this (which is insane) and I can't picture how all those great features could integrate with a scrollable management switch on the fly, but I really think this deserves to be thought about and hopefully added to the feature set sometime down the road.

Maybe a fun side puzzle project for the team?

All my best to all the people making my dream DE come true ❤️

**Comment by jzb** [comment]
Just adding another vote in support of the PaperWM approach - I'm enthused about COSMIC and following its development closely - but I recently tried the PaperWM extension with GNOME and I am in love. It's perfect for my workflow so far. I realize y'all have much on your plates with all the other things you're doing with COSMIC, but this might be a killer feature for its adoption.

**Comment by bprolifcdev** [comment]
Another voice to a growing chorus. PaperWM has revolutionised my workflow. No longer spend most of the day trying to find the window i need to do work in. There is something primally logical about turning left and right to go back to something you were busy with. I use it with ulauncher and the way ulauncher opens to the right of the active window and places the new window to the right of the active window. It feels sooo natural  🥇.

Now lets be realistic. Paperwm is a little buggy, especially with modified shell themes. Ive tried niri but i cant make heads or tales of how to actually use it (I do more that just terminal work) , cosmic plus paperwm type scrolling would be a unique value proposition for the Cosmic DE.

**Comment by mattkang** [comment]
@bprolifcdev Agreed. If you were willing to try niri, I'd suggest [hyprland](https://github.com/hyprwm/Hyprland/) + [hyprscroller](https://github.com/dawsers/hyprscroller). Been using them until we get this in cosmic.

**Comment by Drakulix** [comment]
> We will probably not introduce more window management concepts at this point in development.
> 
> But there have been frequent requests for more options, so we might (long-term) try to provide an api to have external programs add new window management options. Additionally we want to enable changing the compositor, so you could e.g. run niri instead of cosmic-comp in your session to get scrolling window management.

To expand on this: https://github.com/Drakulix/cosmic-ext-extra-sessions/ exists now

**Comment by ygypt** [comment]
While I understand why this may not be a priority for 1.0 release, definitely consider it as a permanent feature to Cosmic in the future.

I believe that scrolling window management is the perfect balance between convenience and usability, especially on laptops where screen space is limited but scrolling is a gesture away.

There is a growing sentiment towards scrolling wms, and while extensions and custom window managers work fine for me, I know plenty of users that are uninterested in that level of configuration but would certainly be enthralled by the feature if it were present. Cosmic could be the first fully fledged DE to natively support it!

**Comment by Prazgreen1** [comment]
honestly, this is a game changer for laptop users due to their very small screen sizes. After all, System76 makes laptops, and I can easily say that tiling is great on laptops, but floating windows in a GNOME like system is better for laptop users as it is easy to navigate between all floating windows. I currently daily-drive a custom GNOME rice with the forefront extension I rely on being PaperWM. This has been a gamechanger as I completely maximize screen real-estate on small displays without at all loosing the convenience of tiling. I would easily say that scrolling tiling is simply the best way to do tiling period. I mean, every scrolling WM works like a normal tiling WM except you get infinite horizontal space with workspaces. Combining a scrolling WM and static workspaces is simply a powerful workflow with no major flaws.

**Comment by s1011205** [comment]
After I came to this thread and checked out PaperWM, I loved that workflow. I actually switched to GNOME primarily to use PaperWM and loved it so much that I don't think that I would switch to a DE where I would not have a good scrollable tiling window manager. I know that I can get that in Gnome, and if nothing changes, that alone would make me use Gnome. Would I rather use Cosmic? Yes! But the workflow with a scrollable tiling window manager is just too good. 

It also has a much lower barrier to entry. I have liked the idea of a tiling window manager in the past, but it would quickly make windows too small and I just did not have the time to configure one and learn how to quickly change size and position of windows to get two specific non-terminal windows where I wanted them. With a scrollable tiling WM, it would be easy to get two windows side by side and taking up around 50% of the main part of the screen without learning more than the two + and - shortcuts for increasing and decreasing width of the window and the shortcuts for moving a window left and right in the horizontal stack. That is enough to have one hooked into the workflow, and then one can slowly learn more of the shortcuts after that.

All that said, I would like to use Cosmic, but if nothing changes, PaperWM on Gnome would have me on Gnome rather than using Cosmic.

**Comment by alicealysia** [comment]
> > We will probably not introduce more window management concepts at this point in development.
> > But there have been frequent requests for more options, so we might (long-term) try to provide an api to have external programs add new window management options. Additionally we want to enable changing the compositor, so you could e.g. run niri instead of cosmic-comp in your session to get scrolling window management.
> 
> To expand on this: https://github.com/Drakulix/cosmic-ext-extra-sessions/ exists now

Been using this with Niri recently, needed to ultimately tweak it so that it runs niri first, then runs the cosmic tools from inside Niri.

As Niri and COSMIC both use much of the same toolset (Smithay, KDL) I think that aiming for better interoperability between the two in future could be a really good idea for both teams.

**Comment by jonathanfhr** [comment]
Also would definitely recommend adding built-in scrollable tling; I was starting to get interested in COSMIC and then discovered PaperWM and am hooked and so have stayed with GNOME. In principle I like the idea of Niri running over COSMIC but don't have the time for the extra setup; if it were built-in that would make COSMIC really attractive.

**Comment by CyberDruga** [comment]
I agree with the comment above, having a scrollable tiling would be awesome, specially because I don't quite like Niri's way of managing workspaces, where each monitor have it's own set of workspaces. I would prefer if they had a way to enable shared workspaces, like PaperWM, but they showed no interest in implementing that, so I've stayed on PaperWM just because I don't feel like changing my workflow 🤷 .

**Comment by Gibbz** [comment]
+1 want to have scrolling as anything else now feels very limited

**Comment by n3ptune-plan3t** [comment]
instead of tiling i feel scrolling + floating would be a better option

**Comment by EA-Wardie** [comment]
I recently started experimenting with Niri, and I have to say it's one of the best window management experiences I've had. I would also like to see something similar in Cosmic.

**Comment by jokeyrhyme** [comment]
I'd love for scrolling to be an available tiling mode, and there's a "mosaic" system that I'd love to see for floating: https://blogs.gnome.org/tbernard/2023/07/26/rethinking-window-management/

is there a future where COSMIC window layouts are pluggable but without it turning into a slow and unstable mess as in other desktop environments?

**Comment by Etienne-TR** [comment]
I believe a built-in solution would align better with COSMIC's philosophy than relying on external tools like niri. It would be more user-friendly and provide consistent default keybindings across tiling modes.

This window management approach is actually quite similar to COSMIC's current tiling mode — the main additions would be implementing infinite horizontal space and adapting the resize logic to apply only to vertically stacked tiles.
Given all the work already done on tiling, I think it's worth considering adding scrolling support. The cost/benefit ratio seems very favorable!

This mode tends to be highly appreciated once discovered — there's a reason why every major DE sees extensions created trying to implement it. But non-native solutions are never fully satisfying. Moreover, scrolling remains hard to discover in most DEs. While it generates enthusiasm when people find it (simple, efficient, and somewhat familiar to users accustomed to mobile interfaces), it could surely be far more popular than it currently is.

Native scrolling tiling would be a huge selling point for COSMIC: not only the only DE offering native tiling, but also the only one with native scrolling!

Thanks to the COSMIC team for all the great work so far — the dedication and quality really show.

**Comment by erialsan** [comment]
If this becomes a reality, I’ll be able to switch from PaperWM + GNOME.
Japanese input—which was the biggest hurdle—is already supported thanks to an update from about a month ago.
I’ve already been testing it on my secondary machine.
If I’m going to use it on my main machine, this is what I want.

**Comment by AnonymousLightbulb** [comment]
I would switch back to cosmic from niri if this was an option!

**Comment by jokeyrhyme** [comment]
the nearest-term solution is: https://github.com/pop-os/cosmic-epoch/issues/322#issuecomment-2628965416

there's also an intriguing (not yet implemented) approach here: https://github.com/pop-os/cosmic-epoch/issues/3402

**Comment by miguel-das** [comment]
For anyone who wants to try scrolling window management in COSMIC, I made a working prototype with GPT-5.6: https://github.com/miguel-das/cosmic_scrolling_prototype

**Comment by kapilkumar9976** [comment]
First of all, thanks to the team for working so hard on a beautiful de. Also, I recommend making this a native option in Cosmic. No other de could match Cosmic's appearance, speed, and feel for me and with this feature, it will be unstoppable and lightning-fast. It will be a killer for Cosmic.

**Comment by r-nl** [comment]
I feel like this is getting quietly sorted as "solved" by a repo that was last updated more then a year ago and isn't really suitable for daily usage.

This is a relatively small push for a window management feature that is rapidly becoming as standard as general tiling, and I do think is worth re-consideration given how major scrolling wms have become in the time since this post was originally made. The landscape has changed, and this is no longer a niche usecase but a fairly major one.

**Comment by boggle** [comment]
Couldn‘t agree more. I‘ve personally moved on to niri and noctalia because
of this and won‘t even reconsider pop-os until this is adressed.

Rob (r-nl) ***@***.***> schrieb am Fr. 18. Sept. 2026 um
19:31:

**Comment by leviport** [comment]
>I feel like this is getting quietly sorted as "solved"

It's not considered solved if the feature request remains open
